feat: compile forward pass llm

llcnt opened the pull request:
## Description
<!-- Provide a brief description of the changes in this PR -->
This PR makes `torch.compile` more suitable to LLMs. We do not compile the whole model (that contains complex dynamic functions), but only the forward pass that computes one token. This change is particularly useful for quantized LLMs: we can obtain speedups by composing `quantization+torch.compile`.
The main code changes are the following:
- add a `causal_lm_logic` in our existing torchcompile file;
- define a `StaticCache` (dynamic KV cache is not compatible with torch compile), and re-define the `generate` function of the LLM with our custom one (max_kv_cache_size is the main bottleneck here: as we can not use dynamic cache, we have to use static cache but with a large enough size so that the user can generate up to max_kv_cache_size tokens);
- create a `utils.py` file where we design the `generate' function. By now we only support 2 input args (input_ids, max_new_tokens). We could add other args (the default generate function from hf contains dozens of args), but it would take time to test all of it.

The results of compatibility tests are the following:
- quanto is not compatible with torchcompile;
- Edit: HQQ is now compatible with torch.compile (see comment below);
- AWQ is not compatible (because AWQ uses a dedicated AWQforcausalLM that wraps the default function generate from hf);
- BnB is compatible when we set fullgraph=False (see #45 );
- GPTQ is compatible ;
- HIGGS (`pruna_pro` feature) is compatible.

## Related Issue
<!-- If this PR addresses an existing issue, please link to it here -->

## Type of Change
<!-- Mark the appropriate option with an "x" (no spaces around the "x") -->
- [ ] Bug fix (non-breaking change which fixes an issue)
- [x] New feature (non-breaking change which adds functionality)
- [ ] Breaking change (fix or feature that would cause existing functionality to not work as expected)
- [ ] This change requires a documentation update

## How Has This Been Tested?
<!-- Describe the tests you ran to verify your changes -->
Latency for running inference with $56$ tokens generated for the prompt "This is a test of this large language model", on a meta-llama/Llama-3.2-1b-Instruct model:
| **Method**      | **GPU**       | **Before Quantization** | **After Quantization** | **After Quantization + Compilation** |
|-----------------|------------------|--------------------------|-------------------------|-------------------------------|
| **BitsAndBytes**| L40S       |        0.83                  |         2.20                |           4.89(fullgraph=false)                    |
|                 | A10G     |           0.94               |           2.41              |             5.24(fullgraph=false)                  |
| **GPTQ**        | L40S       |         0.83                 |          0.76               |             0.75                  |
|                 | A10G     |              0.94            |         0.90                |            0.89                   |
| **HIGGS**       | L40S       |          0.83                |          1.04               |            0.42                   |
|                 | A10G     |           0.94               |          1.23               |            0.48                   |

I have re-tested all methods (w/ and w/o manual cudagraphs after the code refactoring, and the speedups are the same). I have also tested HQQ and HIGGS on a larger llama-3.1-8B model o a single L40S GPU:
| **Method**      |  **Before Quantization** | **After Quantization + Compilation** |
|-----------------|------------------|----------------------------------------------------|
| **HQQ**|       1.59                  |          0.50                   |
| **HIGGS**       |         1.59              |            0.54                  |

I added unit tests on a llama-1B (because opt125 is not compatible)
![image](https://github.com/user-attachments/assets/9cad2ea5-c213-4ca1-ac43-e54efc9cea9f)

Experiments can be reproduced with this [notebook](https://drive.google.com/file/d/1GaQTVdWuF8RcjhP2n-ZcjuqomQogmP9B/view?usp=sharing).

## Checklist
<!-- Mark items with "x" (no spaces around the "x") -->
- [x] My code follows the style guidelines of this project
- [x] I have performed a self-review of my own code
- [x] I have commented my code, particularly in hard-to-understand areas
- [x] I have made corresponding changes to the documentation
- [x] I have added tests that prove my fix is effective or that my feature works
- [x] New and existing unit tests pass locally with my changes

## Additional Notes
<!-- Add any other information about the PR here -->
- opt models are not compatible (because they dont have a 'cached_postion' arg), so an exception check was created to avoid compiling an opt model that is quantized (IMO it is ok as opt models are currently outdated)

sharpenb commented on src/pruna/algorithms/compilation/utils.py:
```diff
@@ -0,0 +1,151 @@
+
+
+@torch.compile
+def sample(logits, temperature: float = 1.0, top_k: Optional[int] = None):
```
I see now that it is restricted to one sampling strategy while there could be multiple ones (see [here](https://huggingface.co/docs/transformers//generation_strategies)). I wuold suggest ot check if we cna inherit from theirs and use the same trick to compile all the options.

llcnt replied:
Yes we can :) But as I mentioned in the main comment, we would have to write by ourself each if statement (corresponding to each generation strategy; as it is done in`transformers` [here](https://github.com/huggingface/transformers/blob/7ecc5b88c0328aea91a3c9f8763f56b3b1e26767/src/transformers/generation/utils.py#L2101) ) and to test them all. Should we spend time on this now, or postpone this work for a future PR? (I would be in favor of the latter for keeping this PR as simple as possible ahah ;) )

sharpenb replied:
Makes sense. I am still wondering if there is no function in hf that we can simply import/inherit and does the similar/same things? Now we reimplement the function to the lowest details.

llcnt replied:
Yes I guess I can see what you have in mind. :) My initial plan was to get the "smallest" function to be compiled so that we can after "put it back into the model", and have something like "function_before_generate = our_custom_compiled_function", so that we can blindly use the default `generate()` function afterwards. Unfortunately it is not possible: we compile the function that gets as input the (previous) tokens, and generate a new token. The limitation comes from the fact that `transformers` does not re-uses any function between the different sampling strategies. Currently we have done `model.generate = our_custom_compiled_generate`, what we could do as a future work is to do :
- `model._sample = our_custom_compiled_generate_sample` (this is the function we currently "overwrite" by [default](https://github.com/huggingface/transformers/blob/e5ac23081ec4021818a21d7442d396f31de8c30c/src/transformers/generation/utils.py#L2452) when we do the model.generate replacement);
-  `model._beam_search= our_custom_compiled_generate_beam`
-  `model._assisted_decoding= our_custom_compiled_generate_assisted`
-  `model._dola_decoding= our_custom_compiled_generate_dola`
-  `model._contrastive_search= our_custom_compiled_generate_contrastive`
- `model._group_beam_search= our_custom_compiled_generate_groupbeam`
- `model._constrained_beam_search= our_custom_compiled_generate_constrainedbeam`

sharpenb replied:
Thanks for the explanation! I think that this implementation is nice enough for this PR. For a future PR that will be able to compile any specified function I feel that another implementation would be possible. The implementation could make a new function that takes as input any function that would need to be compiled, create a replacement compiled function from it without detailed reimplementation (e.g. by wrapping the input function and compiling it) and replace it in the model. There would be two advantage from this implementation: easier maintenance through no code boilerplate, and capacity to compile functions at different levels :)

sharpenb commented on src/pruna/algorithms/compilation/utils.py:
```diff
@@ -0,0 +1,151 @@
+        -------
+            torch.Tensor: The generated ids
+        """
+        batch_size, seq_length = input_ids.shape
```
This function would be enfit from commenting to explain the logic. E.g. it is unclear where the past_kv is updated. It looks like when generating a new token the kv for this token is not update in the past kv.

sharpenb replied:
I feel now that it is a bit too many comments. Could we siply say (1) init generated token id and their position, (2) sample/decode tokens. One question is also why we cannot use compiled decoding to generate the first token?

llcnt replied:
I completely refactor so that now we have a HFGenerator and some functions I tried to comment :)

sharpenb reviewed (changes requested):
Thanks for the PR! The PR description with the benchmark is very promising :) In a future PR, I feel that we would benefit to have more llm tests on small llms (beyond opt).

llcnt commented on src/pruna/algorithms/compilation/torch_compile.py:
```diff
@@ -45,7 +49,7 @@ class TorchCompileCompiler(PrunaCompiler):
     dataset_required = False
     compatible_algorithms = dict(
-        quantizer=["half", "hqq_diffusers", "diffusers_int8"],
+        quantizer=["half", "hqq_diffusers", "diffusers_int8", "gptq", "llm_int8"],
```
No we dont need, compatibilities are un-directed in between pruna and pruna_pro ;)

llcnt commented on tests/fixtures.py:
```diff
@@ -107,7 +107,15 @@ def get_automodel_transformers(model_id: str) -> tuple[Any, SmashConfig]:
 
-    smash_config.add_data("WikiText")
+    # quantizers that require calibration will throw an error if the tokenizer does not have a pad token
+    tokenizer_for_calibration = AutoTokenizer.from_pretrained(model_id)
```
There is no need to change this if we test quantizer that requires calibration on opt125 (that is the case for our previous tests). However if you test a llama, it will fail beacause the tokenizer does not have a pad token (we did not have any trouble before because when we were testing llama, eg on HQQ, we did not use calibration, hence no tokenizer) :)
- default llama1B tokenizer does not have a pad token;
- we can delete this limit, but the unit will be slower (considering that we test here a llama1b model and not an opt125)

llcnt replied:
Edit: I integrated the fix from begum when rebasing, but I still need to limit the dataset size (otherwise it kills my cpu)

llcnt commented:
Edit2: I have added support for HQQ !
- We don't use our default forward compilation, but the one provide in HQQ where cuda graphs are managed manually;
- during my experiments I experienced cuda errors when using a cache_size for HQQ+compile < 400. I also experienced some bugs when the number of generated tokens is small (<100)
- I also experienced errors (reported [here](https://github.com/mobiusml/hqq/issues/157)) when using hqq-quantized from Automodel from hf, or when using a model in torch.float16: we need torch.bfloat16 and a model that uses AutoHQQ;
- The generator provided in HQQ directly converts input/output with the tokenizer corresponding to the given LLM. So I added some lines to convert this when necessary so that the user can follow the same pipeline `model.generate(input_tokenized, max_new_tokens)`, as it is the case in `transformers`.

I have also modified the hqq.py file because some users (including me ahah) have experienced some bugs (impossible to move the model with .to):
- I removed the `try:` statement that tries to use .to to the hqq-quantized model: it is un-valid by nature;
- I added a torch_dtype (in Automodel call) and a compute_dtype (in AutoHQQ call) arg to be able to use torch.bfloat16;
- I exchanged the AutoHQQ and AutoModel in the try statement so that, by default, we first try to apply AutoHQQ now. I did this because, for some reason, `prepare_for_infererence` did not change the HQQmAtmul when the model is quantized with AutoModel from `transformers`;
- I re-launched the unit tests on HQQ to be sure my change did not break anything and there are successful.

sharpenb commented on src/pruna/algorithms/compilation/torch_compile.py:
```diff
@@ -241,6 +265,67 @@ def unet_transformer_pipeline_logic(model: Any, smash_config: SmashConfigPrefixW
+        past_kv = StaticCache(
+            model.config,
+            smash_config["batch_size"],
+            smash_config["max_kv_cache_size"],
```
This parameter simpply be the context length?

sharpenb replied:
Makes sense. Since LLM usually come with a `max_context_length` variable, I could see that we remove this as hyperparameter and simply set it to this value. If a user wants to restrict the cache size, they can simply cahnge the value of this vairable in the LLM they give as input :)

sharpenb commented on src/pruna/algorithms/compilation/torch_compile.py:
```diff
@@ -241,6 +265,67 @@ def unet_transformer_pipeline_logic(model: Any, smash_config: SmashConfigPrefixW
+                do_sample=True,
+                compile="partial",
+                compile_options={"mode": smash_config["mode"], "fullgraph": smash_config["fullgraph"]},
+            ).enable_cuda_graph()
```
If I understand correctly, the compilation happens here now. When looking at [this](https://github.com/mobiusml/hqq/blob/1f052eb5a0aab0572d380d48b708ae1c74936d23/hqq/utils/generation_hf.py#L112), it looks like there is quite some redundant code with what we have in utils. Could we simply have our own equivalent of HFGenerator class that we use for all quantization methods? It would avoid this if HQQ and make maintenance easier.

llcnt replied:
I refactored the code to follow your idea ;) we have now a single generator class :)

sharpenb replied:
Amazing, thanks! Did you rerun the quick benchmark to see that it does not change the speedup? It should not change anything but could be nice to check a couple of the table rows :)

sharpenb replied:
Could we also add the HQQ number to the table? :)

sharpenb replied:
Thanks a lot for it, that is cool :)

sharpenb reviewed (changes requested):
Thanks for the nice udpate! I left some comments on the modified files. I feel that some code factorization is possible :)

sharpenb commented on src/pruna/algorithms/compilation/torch_compile.py:
```diff
@@ -241,6 +271,49 @@ def unet_transformer_pipeline_logic(model: Any, smash_config: SmashConfigPrefixW
+            compile_fullgraph=smash_config["fullgraph"],
+            batch_size=smash_config["batch_size"],
+        )
+    # If we are using max-autotune-no-cudagraphs, we need to handle the cudagraphs manually.
```
This part is a bit unclear. It feels that by enabling the cuda graph we would do sthg similar to `max-autotune` (also when reading the [docs](https://pytorch.org/docs/stable/generated/torch.compile.html)). Could you elaborate on what is the benefit of this no cudagraphs mode? :)

llcnt replied:
I was not able (I tested on qwen7b, llama1B and llama8B) to spot any errors when using "max-autotune". However, some models are supposed to not be able with torch_compile automatic cuda graphs. in this case, we would have to manage them manually. I tested with and without these manual cudagraphs, and the speed results are exactly the same. But as the additional code lines for manual cudagraphs are minimal I would be in favor to keep them ;)

sharpenb commented on src/pruna/algorithms/compilation/utils.py:
```diff
@@ -0,0 +1,530 @@
+
+        if cleanup:
+            del self.inputs, self.generated_ids, self.cache_position
+            torch.cuda.empty_cache()
```
We might need as well `gc.collect()` to finish the cleaning.

llcnt replied:
Indeed! I now use our `safe_memory_cleanup` function from the utils ;)

sharpenb commented on src/pruna/algorithms/compilation/utils.py:
```diff
@@ -0,0 +1,530 @@
+
+        Parameters
+        ----------
+        next_token : torch.Tensor
```
That sounds a bit unatural to me that the current token is called next_token. Also l.530 looks to input here prefill() which intuitvely gives the feeling that these are the previous tokens. We can probably make these variables name more clear :)

llcnt replied:
Agrre ;) I cleaned all this current vs next token, to make the code more consistent

sharpenb commented on src/pruna/algorithms/compilation/utils.py:
```diff
@@ -0,0 +1,530 @@
+        self.generated_ids[:, self.cache_position] = next_token.int()
+        return next_token
+
+    def gen_next_token(self, next_token: torch.Tensor) -> torch.Tensor:
```
It looks like we could merge gen_next_token and gen_next_token_raw :)

llcnt replied:
Done :)

sharpenb commented on src/pruna/algorithms/compilation/utils.py:
```diff
@@ -0,0 +1,530 @@
+from transformers.cache_utils import StaticCache
+
+
+class HFGenerator:
```
It is not clear that we should call it HF Generator. Is the function specialized to CausalLM or transformers? IN general, it feels that this class/code is a good base to adapt but we can probably think how to clean it :)

llcnt replied:
I see your point. We indeed re-write the sampling process of the LLM, and this is not depending on `transformers`. However, all the implementation is currently tied to the management of the `StaticCache`, which is a 'HF' feature, and is not working (to best of my knowledge) with other LLM implementations

sharpenb replied:
Makes sense. It looks to me that `StaticCache` is imported from transformers. If this class is tied to some package/model format, I would then suggest to rename it to represent this specification for now e.g. `TransformersGenerator`. This is not blocking the PR but would be more informative and the current implementation constraints.

llcnt replied:
I changed the class :)

sharpenb commented on src/pruna/algorithms/compilation/utils.py:
```diff
@@ -0,0 +1,530 @@
+        self.compile_mode = compile_mode
+        self.compile_fullgraph = compile_fullgraph
+        self.batch_size = batch_size
+        self.decode_one_token = self.decode_one_token_sampled
```
Could we merge the functions decode_on_token and decode_one_token_sampled?

llcnt replied:
Done :)

sharpenb commented on src/pruna/algorithms/compilation/utils.py:
```diff
@@ -0,0 +1,530 @@
+
+        self.setup_cache()
+
+        self.is_compiled = False
```
I would suggest to replace l103-105 with
```
self.is_compiled=False
self.decode_one_token = torch.compile(decode_one_token, mode=self.compile_mode, fullgraph=self.compile_fullgraph)
self.is_compiled=True
```
It would make the overall code much shorter :)

llcnt replied:
Yes! I have also deleted the is_compiled :)

sharpenb commented on src/pruna/algorithms/compilation/utils.py:
```diff
@@ -0,0 +1,530 @@
+    def enable_cuda_graph(
+        self,
+        iters: int = 2,
+        prompt_tokenized: list[int] = [596, 8830, 315, 6913, 19476, 11, 1778, 439, 279, 12939],
```
This is the random prompt, right?

llcnt replied:
Yes :)

sharpenb commented on src/pruna/algorithms/compilation/utils.py:
```diff
@@ -0,0 +1,530 @@
+            torch.tensor(prompt_tokenized, device=self.model.device).unsqueeze(0),
+            max_new_tokens=max_kv_cache_size
+        )
+        for _ in range(iters):
```
Why do we need mulitple iterations? It looks like it might be not needed to capture the cuda graph.

llcnt replied:
When dealing with manual cudagraphs, we need to run a static input without graph first (of shape 1), and then run multiple iterations of graph captures, as derailed in [this thread](https://github.com/huggingface/transformers/issues/30351)

sharpenb commented on src/pruna/algorithms/compilation/utils.py:
```diff
@@ -0,0 +1,530 @@
+                max_new_tokens=max_kv_cache_size
+            )
+
+    def gen_next_token_withgraph(self, next_token: torch.Tensor) -> torch.Tensor:
```
I think that all these cuda_graph function are done to support max auto tune nograph mode. It could be nice to discuss when is this mode particularly useful (see my comment in torch compile) :)

llcnt replied:
Yes I technically experienced no difference, but could be some, see my answer above ;)

sharpenb commented on src/pruna/algorithms/compilation/utils.py:
```diff
@@ -0,0 +1,530 @@
+        """
+        return self.gen_next_token_raw(next_token)
+
+    def enable_cuda_graph_(self) -> None:
```
I feel that we could merhe `enable_cuda_graph_` and `enable_cuda_graph`

llcnt replied:
done !

sharpenb reviewed (changes requested):
Thanks for the udpate! It is pretty nice to factorize the compilation code for multiple quantization methods. While HFGenerator class is a nice base, I believ that we can clean the code so that it will be shorter and easier to maintain. Happy to dicsuss :)

sharpenb commented on src/pruna/algorithms/compilation/utils.py:
```diff
@@ -0,0 +1,530 @@
+            This method modifies the internal state of the generator to use CUDA graphs
+            but does not return any value.
+        """
+        _ = self.generate(
```
I was checking the thread and it mostly mentions forward and warmup after compilation.
![image](https://github.com/user-attachments/assets/19907f1c-6099-412d-b307-8c21bde73339)

sharpenb replied:
It is also look that we use the same random prompt for all the passes before/after the compilation. Out of curiosity, what happens if we do only one forward pass on this single prompt? 
![image](https://github.com/user-attachments/assets/51820f25-2af1-47f0-b1e9-d53c70289700)

llcnt replied:
Yes, this is my bad, here is the full explanation:
- if you run directly the model.generate with the graph capture, you will get the following error `InternalTorchDynamoError: RuntimeError: Cannot call CUDAGeneratorImpl::current_seed during CUDA graph capture.`, you can reproduce this error by running this [notebook](https://drive.google.com/file/d/1HkcOS3O-RHJJrefDv_LXkHlsmDgHumEt/view?usp=sharing) and just comment out the line `_ = self.generate` without cuda graphs (I left the error in the cell output so that you can see ;) )
- my explanation (but tbh I am not sure, I dont know the subtilities about cuda graphs) is that we have to run a 'normal' generate before activating manually the cuda graphs capture, maybe to fill the KV cache with some values ? I guess this is what they mean by 'Compile with inputs of size [batch_size, 1]' before doing the warmup? What I still dont understand is that this step is not compiled
- what is implemented [here](https://github.com/mobiusml/hqq/blob/1c47dfc2e58bac717087cd215cf45d1d1afd3eea/hqq/utils/generation_hf.py#L354), also do the call without enabling the cuda graph before. Maybe because they experienced the same error while directly capturing the cuda graph?

sharpenb commented on src/pruna/algorithms/compilation/utils.py:
```diff
@@ -0,0 +1,479 @@
+        self.model.generation_config.cache_implementation = "static"
+        self.model.config.use_cache = True
+
+    def multinomial_sample_one_no_sync(self, probs_sort: torch.Tensor) -> torch.Tensor:
```
It is good for this PR to have our own code. However, it still feels that we could reimport some of their functions and compile it e.g. the list of smapling methods defined [here](https://github.com/huggingface/transformers/blob/4602059aaee7b075ca51c21941ab2d27980bef7c/src/transformers/generation/utils.py#L322). Again, that is good for now but if it is possible that would be much ncier to import all samplers and so that they can all be compiled :)

llcnt replied:
Yes maybe some functions (like this multinomial_sample, or the logits compute) can be re-used across other sampling schemes. However, we will still need to define and compile a given function for each sampling strategy, because currently in `transformers`, the generate function is made of a long stack of `if` statements (one for each sampling strategy), and each sampling has its own logic (it seems that only the model.forward is re-used across sampling strategies): 
![image](https://github.com/user-attachments/assets/d874ec70-ea2e-4384-bc17-af7ce55efb63)

sharpenb reviewed (approved):
Thanks for the udpate! I think that we converged to sthg that we can merge :) As discussed, there would be a list of updates that can be done to optimize and extend the generality of the code :)

johnrachwan123 reviewed (approved):
Thanks a lot for the great work! Let's merge 🚀
